### PR TITLE
Mat out

### PR DESCRIPTION
**Pull request by Takishima**

### CODE REVIEW

**Comment by Takishima** [comment]
So @XYShe ,

As you might have noticed I have rewritten your code quite as I was reviewing it since some of the changes were relatively easy to implement.

What I would like you to do now is go through my changes and see if I have messed up something somewhere. I do not think so since the tests are still passing, but just in case I have misunderstood some of your hypotheses and something could go wrong thereafter.

Also, small optional question: do you know how we might want to handle the case where we deallocate a qubit and how we might need to modify the existing unitary in that case?

Similarly to what the simulator is doing, if we know the classical value of the qubit, could we extract the part of the unitary with the corresponding value?
Could we for example use something like this?

```python
# Assuming that we want to delete the qubit at position 3
mask = _qidmask([0, 1, 2 4], [], [], 5)

# Could we use either mask[0] or mask[1] to extract the matrix from the unitary stored in the UnitarySimulator?
```

**Comment by XYShe** [comment]
> So @XYShe ,
> 
> As you might have noticed I have rewritten your code quite as I was reviewing it since some of the changes were relatively easy to implement.
> 
> What I would like you to do now is go through my changes and see if I have messed up something somewhere. I do not think so since the tests are still passing, but just in case I have misunderstood some of your hypotheses and something could go wrong thereafter.
> 
> Also, small optional question: do you know how we might want to handle the case where we deallocate a qubit and how we might need to modify the existing unitary in that case?
> 
> Similarly to what the simulator is doing, if we know the classical value of the qubit, could we extract the part of the unitary with the corresponding value?
> Could we for example use something like this?
> 
> ```python
> # Assuming that we want to delete the qubit at position 3
> mask = _qidmask([0, 1, 2 4], [], [], 4)
> 
> # Could we use either mask[0] or mask[1] to extract the matrix from the unitary stored in the UnitarySimulator?
> ```

Hi Damien, thank you very much for the changes! I think everything looks right. I'll implement an additional check for the off-control (after merging with ctrl_generalize) so the decomposition rule is invoked for this backend. Afterwards, I believe the ctrl_state is no longer needed to compute the unitary.

Also, I believe that is the correct way to deallocate a qubit or even multiple qubits at the same time: _qidmask([0,1,2,4],[],[],5). Similarly, it should be quite similar for measurement except maybe we can use the ctrl_qubit and ctrl_state to filter out the right value. However, I might need to test to make sure using measured value within the backend doesn't cause any bugs.

File: projectq/backends/_unitary.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    """
+    mask_list = []
+    perms = np.array([x[::-1] for x in itertools.product("01", repeat=n_qubits)]).astype(int)
+    all_ids = np.array(range(n_qubits))
+    irel_ids = np.delete(all_ids, control_ids + target_ids)
+    m = len(irel_ids)
+
+    if len(control_ids) > 0:
+        cstate = np.array(list(control_state)).astype(int)
+        cmask = np.where(np.all(perms[:, control_ids] == cstate, axis=1))
```
Comment (Takishima) [inline projectq/backends/_unitary.py]: I forgot to ask you, if we anyway do not support negatively controlled qubits (with `state == 0`) what is the point of using the `state` argument in the first place?

Comment (XYShe) [reply projectq/backends/_unitary.py]: It was an attempt to fix a bug but turns out that was not the cause: all incoming ctrl_state should be 111... due to decomp rule. This is removed.

Comment (Takishima) [reply projectq/backends/_unitary.py]: Ok.

**Comment by Takishima** [comment]
GitHub seems to be listing the wrong commits in this PR. I will close it and then re-open a new one to hopefully fix this.

**Pull request closed without merging**